[WIP] Add truncation support to `pg_query_summary()`

duckinator opened the pull request:
TODO:
- [x] import `list_sort()`
- [x] import `pg_mbstrlen()`
- [x] import `pg_mbcharcliplen()` and `cliplen()`

duckinator committed:
Re-enable existing code for truncated queries, port truncation tests from Rust to C, add TEST_ASSERT_NULL().

duckinator commented:
Tests _should_ be failing on `test/summary_truncate`, right now.

duckinator committed:
regenerate protobuf stuff

duckinator committed:
split pg_query_summary_truncate() to its own file.

duckinator committed:
[WIP] [truncate] Implement tree walking code.

duckinator committed:
[truncate] Track depth

duckinator committed:
[truncate] implement dummy_{column,target,select,insert,update}

duckinator committed:
[truncate] Add pg_query_deparse_node()/_deparse()/_deparse_opts(); implement first pass of *_len() functions.

duckinator committed:
[truncate] Add crude, non-UTF8-compliant implementation of truncate_str().

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,409 @@
+	stmt->targetList = targetList;
+	stmt->whereClause = NULL;
+	stmt->returningList = NULL;
+	stmt->withClause = NULL;
```
It appears that `makeNode` internally calls `palloc0` ([Postgres source](https://github.com/postgres/postgres/blob/master/src/include/nodes/nodes.h#L155)), so I don't think we need to initialize anything where the zero value is acceptable (i.e. all these ` = NULL;` lines can be omitted).

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,409 @@
+{
+	// FIXME: This WILL blow up for multi-byte UTF-8 characters, since it's going
+	// by byte instead of character. I think C has functions to deal with this,
+	// I just haven't learned about them yet. -duckinator
```
It looks like we can use `pg_mbcliplen` for this, which lets you specify the number of bytes you want to limit to, but respects multi-byte boundaries: https://github.com/postgres/postgres/blob/master/src/backend/utils/mb/mbutils.c#L1084

Alternatively, I wonder if we should treat the truncation limit as number of multi-byte characters, not number of bytes. In that case, `pg_mbcharcliplen` could be used: https://github.com/postgres/postgres/blob/master/src/backend/utils/mb/mbutils.c#L1126 (and we should make sure to use `pg_mbstrlen` to calculate the length of strings)

duckinator replied:
It's now truncating based on the number of characters, since that appears to be what the original intent was.

The functions that are temporarily copied and need to be imported have been added to the TODO list for this PR.

duckinator committed:
[truncate] temporarily include list_sort(); sort list of possible truncations

duckinator committed:
[truncate] Avoid explicitly setting fields to NULL when using makeNode(), since it uses palloc0().

duckinator committed:
[truncate] when bailing early, be sure to free unused things.

duckinator committed:
[truncate] Remove more unnecessary NULL assignments.

duckinator committed:
[truncate] Split out a test_run_impl() as a common implementation for test_run() and test_run_with_mcxt().

duckinator committed:
Add TEST_FAIL_FAST, which causes test_run()-and-friends stop after the first test failure.

duckinator committed:
[truncate] improve error reporting (even if it's not handled well yet.)

duckinator committed:
[truncate] Wrap things in a RawStmt if needed.

duckinator committed:
Have deparseSelectStmt() bail early if stmt == NULL.

duckinator committed:
[truncate] Add basic structure of apply_truncations()

duckinator committed:
[truncation] minor refactoring; remove now-unneeded printf statements.

duckinator committed:
[truncate] Almost have a successful truncation

duckinator committed:
improve error message for TEST_ASSERT_STR_EQUAL

duckinator committed:
[truncate] remove unneeded print statements

duckinator committed:
[truncate] add missing return

duckinator committed:
[truncate] handle "…" -> ... expansion and SELECT ... AS ... -> SELECT ... kludge.

duckinator committed:
[truncate] stop applying truncations if we meet the length limit

duckinator committed:
[truncate] If the untruncated version is short enough, return it immediately.

duckinator committed:
[truncate] implement other branches of apply_truncations()

duckinator committed:
[truncate] fix fallback crude truncation.

duckinator committed:
[truncate] apply fallback truncation the latest attempt, not the original version

duckinator committed:
[truncate] change some values in dummy_select() from 1 to 0, which fixes multiple tests, then remove everything that's 0 or equivalent.

duckinator committed:
[truncate] select_values_lists_len() generates things like "VALUES (...)" not "SELECT ..."

duckinator committed:
[truncate] If an item is <=3 characters long, refuse to remove it when truncating.

duckinator committed:
[truncate] remove the test i added, since it seems to be redundant with one i missed previously

duckinator committed:
[truncate] truncate strings to a specific number of characters, not bytes.

duckinator committed:
[truncate] Rename pg_query_deparse to pg_query_deparse_list, and add a pg_query_deparse that uses the _node and _list variants as appropriate.

duckinator committed:
make deparseStmt() support CommonTableExpr.

duckinator committed:
[truncate] Improve error reporting in truncation-related code.

duckinator committed:
[truncate] make summary_truncation_options() properly handle CTEs.

duckinator committed:
[truncation] When encountering a CTE, remove possible truncations that involve the same node.

duckinator committed:
[truncation] Use memmove() instead of strcpy() to avoid data corruption.

lfittl commented on src/pg_query_deparse.c:
```diff
@@ -10,6 +10,81 @@
+	return pg_query_deparse_opts(stmts, opts);
+}
+
+PgQueryDeparseResult pg_query_deparse_opts(List *stmts, PostgresDeparseOpts opts)
```
Thanks for adding this!

I think we could consider refactoring `pg_query_deparse_protobuf_opts` to just call this (after doing pg_query_protobuf_to_nodes), but we can do that in a follow-up PR as well.

lfittl commented on src/pg_query_deparse.c:
```diff
@@ -10,6 +10,81 @@
+	return pg_query_deparse_list(list_make1(node));
+}
+
+PgQueryDeparseResult pg_query_deparse_list(List *stmts)
```
I'd rename this as `pg_query_deparse_stmt_list` to make it clear this must be passed a list of statements, not something else (e.g. a list of expressions is not valid).

lfittl commented on src/pg_query_deparse.c:
```diff
@@ -10,6 +10,81 @@
 
 #include "protobuf/pg_query.pb-c.h"
 
+PgQueryDeparseResult pg_query_deparse(Node *node)
```
I think it'd be better to leave the decision whether its a node (that potentially needs to be wrapped) or a list of RawStmt nodes to the caller, so if we introduce support for deparsing expressions (instead of raw statements) in the future, we don't have a generic function name making the API confusing to understand.

lfittl commented on src/pg_query_deparse.c:
```diff
@@ -10,6 +10,81 @@
+		return pg_query_deparse_node(node);
+}
+
+PgQueryDeparseResult pg_query_deparse_node(Node *node)
```
As discussed, maybe naming this as `pg_query_deparse_stmt` is better here, since I think the node must still be a valid statement (to be wrapped in a `RawStmt`), as compared to e.g. an A_Expr node.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+}
+
+static bool
+summary_truncation_options(Node *node, TruncationState *state)
```
Maybe name this `generate_possible_truncations`?

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+	{
+		case T_RawStmt:
+			// For some reason our statement has been trapped in a RawStmt.
+			// Free them.
```
I'd probably skip that comment :-)

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+					add_truncation(state, TRUNCATION_COLS, node, cols_len(stmt->cols));
+
+				break;
+			}
```
We may want to add support `T_MergeStmt` here as well (which didn't exist when the truncation code was first written, so this is an existing bug, but should be easy to address here)

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+				break;
+			}
+
+		case T_InsertStmt:
```
This is an existing issue, but maybe (nitpicking my past self here I guess?) the InsertStmt fits better before the UpdateStmt, so the order is Insert/Update/Merge/Delete/Copy.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+				CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+
+				if (stmt->ctequery != NULL) {
+					PgQueryDeparseResult result = pg_query_deparse(node);
```
I think this needs to be `(Node *) stmt->ctequery`, so we're actually passing something that is a statement.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+						return false;
+					}
+
+					int32_t length = strlen(result.query);
```
I'd probably inline this in the function call below, to match the other `_len` calls (though this is a different kind of function being called, I don't think putting it in a variable is necessary)

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+	PgQueryDeparseResult result = pg_query_deparse_node(dummy_select(nodes, NULL, NULL));
+
+	if (result.error) {
+		print_pg_query_error(result.error, __func__);
```
We could investigate if we can make `elog(ERROR, ..)` work here, so the whole truncation fails, instead of making it not truncate that portion. Because we work with a known input (a valid parse tree), we can require that the deparser always succeeds, and consider it a bug if it doesn't.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+
+	state->depth++;
+
+	return raw_expression_tree_walker(node, summary_truncation_options, (void *) state);
```
As discussed, this won't correctly behave when there is sibling nodes that the walker gets called for, so we'd either need to pass this to the function as a separate argument (seems challenging to implement), or restore the previous value after having descended down, like this:

```suggestion
	int old_depth = state->depth;
	state->depth++;

	bool result = raw_expression_tree_walker(node, summary_truncation_options, (void *) state);

	// Restore old depth value since the current node (or its parents) may have sibling elements
	state->depth = old_depth;

	return result;
```

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+	// pattern is a double quote, ellipses, and another double quote.
+	char pattern[5] = {'"', 0xE2, 0x80, 0xA6, '"'};
+
+	for (size_t i = 0; i < strlen(str); i++) {
```
I think we'd need to subtract the pattern length from strlen here, so we don't overrun when the pattern would no longer fit into the remaining string.

duckinator replied:
I looked into this and as best I can tell it's guaranteed to never overrun just because of how `memcmp` compares things.

The pattern is `"…"`, and it's 5 bytes. The edge case we're worried about is it being at the end of the string.

These are the only outcomes I could manage to create:

1. it makes 0 or more successful matches, then a null byte, and `memcmp(str + i, pattern, 5) == 0` returns a non-zero result immediately after encountering the null byte.
2. it has a syntax error, and never gets to `expand_ellipses()`.

lfittl replied:
Good point, I didn't think about the null byte comparison. Agreed this should be safe as is :)

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+	return raw_expression_tree_walker(node, summary_truncation_options, (void *) state);
+}
+
+static int cmp_possible_truncation_depth(const ListCell *a, const ListCell *b)
```
As discussed, the intent here is to "Truncate the deepest possible truncation that is the longest first" (see [Ruby implementation](https://github.com/pganalyze/pg_query/blob/main/lib/pg_query/truncate.rb#L16)), so we're missing the length of the truncation in this comparison. I think we can probably just rename this function `cmp_possible_truncation_depth` and add length to it (I think you'd just check depth first, return early if its different, and if its the same then compare length, making sure that longer things sort first, i.e. the opposite logic we have for depth).

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+	if (strlen(result.query) <= truncation_limit) {
+		summary->truncated_query = output;
+		return NULL;
+	}
```
We could do this before generating truncations to save on the generation if the query is already short enough.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+			CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+			Node *old = stmt->ctequery;
+			stmt->ctequery = dummy_select(NULL, (Node *) dummy_column(), NULL);
+			if (old) {
```
To capture my understanding: This logic (that came from Rust originally, and is not present in the Ruby implementation) helps us not revisit the statement contained (e.g. `SelectStmt`), because we already removed the whole CTE anyway. Because it is valid to update a removed node (since we do not free it, we just remove it from the tree), this should not affect correctness, but rather it lets us potentially skip an unnecessary deparse.

However, I think because we go depth first, this may never be utilized, so its potentially worth dropping here.

lfittl replied:
Also, if we remove this, we could replace the iteration logic with a foreach, since we don't actually have to modify the list as we iterate over it.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+		Node *node = truncation->node;
+		enum TruncationAttr attr = truncation->attr;
+
+		if (truncation->length <= 3) {
```
It might be worth avoiding creating these truncations in the first place.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+		} else if (IsA(node, SelectStmt) && attr == TRUNCATION_TARGET_LIST) {
+			SelectStmt *stmt = castNode(SelectStmt, node);
+			stmt->targetList = list_make1(dummy_target());
+			printf("Select/targetList\n");
```
Per our discussion, these were just for debugging and will go away :)

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,673 @@
+			IndexStmt *stmt = castNode(IndexStmt, node);
+			stmt->whereClause = (Node *) dummy_column();
+			printf("Index/whereClause\n");
+		}
```
If we add support for MergeStmt, we should of course also add it here.

lfittl commented on src/postgres_deparse.c:
```diff
@@ -11869,6 +11872,9 @@ static void deparseStmt(DeparseState *state, Node *node)
 		case T_CommentStmt:
 			deparseCommentStmt(state, castNode(CommentStmt, node));
 			break;
+		case T_CommonTableExpr:
```
This should not be needed once we fix the case where we pass `node` instead of `stmt->ctequery` to deparse (when generating truncations).

lfittl commented on src/postgres_deparse.c:
```diff
@@ -2730,6 +2730,9 @@ static void deparseUtilityOptionList(DeparseState *state, List *options)
 
 static void deparseSelectStmt(DeparseState *state, SelectStmt *stmt, DeparseNodeContext context)
 {
+	if (stmt == NULL)
```
Its not clear to me why that would happen, might be worth removing and seeing what needs this (if anything).

lfittl commented on test/framework/main.c:
```diff
@@ -164,18 +178,39 @@ TEST_ASSERT_LIST_EQUAL_impl(TestState * test_state, char *actual_str, List *actu
+// Given a list of tests, cleanup function, and use_mctx,
+// run tests and return the results.
+//
+// If `use_mctx` is true, a Postgres MemoryContex is enter before the test is run,
```
```suggestion
// If `use_mctx` is true, a Postgres MemoryContex is entered before the test is run,
```

duckinator committed:
[truncate] Remove pg_query_deparse(), rename pg_query_deparse_list() to pg_query_deparse_stmt_list(), throw error if passing the wrong node type to either.

duckinator committed:
[truncate] remove unneeded comment

duckinator committed:
[truncate] Rename summary_truncation_options to generate_possible_truncations for clarity.

duckinator committed:
[truncate] for CTE Query truncations, call deparse on the actual ctequery, and inline the strlen call.

duckinator committed:
[truncation] Rename pg_query_deparse_node to pg_query_deparse_stmt.

duckinator committed:
[truncation] Remove print_pg_query_error() and switch to elog().

duckinator committed:
[truncate] If the query doesn't need to be truncated bail immediately, instead of after generating possible truncations.

duckinator committed:
[truncate] Handle state->depth properly.

duckinator committed:
[truncate] Fix cmp_possible_truncations, which also fixes a test!

duckinator committed:
[truncation] Don't handle CommonTableExpr in deparseStmt (that was only needed because of a bug elsewhere).

duckinator committed:
[truncate] Revert addition of NULL check that was only needed because of a bug elsehwere.

duckinator committed:
[truncate] Fix typos in comment.

duckinator committed:
[truncate] Remove weird CTE section that doesn't seem to affect anything.

duckinator committed:
[truncate] Switch to foreach, since we no longer modify the list during the loop.

duckinator committed:
[truncate] Move IndexStmt before UpdateStmt; add beginnings of work for MergeStmt.

duckinator committed:
[truncate] Fix expand_ellipses() and remove_unneeded_as().

duckinator committed:
[truncate] Fix memory leaks.

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,684 @@
+			size_t len = strlen(str + i + 5);
+			memcpy(str + i, "...", 3);
+			memmove(str + i + 3, str + i + 5, len + 1);
+		}
```
```suggestion
size_t pattern_len = 5;
char* replace = "...";
size_t replace_len = strlen(replace);
...
		if (memcmp(str + i, pattern, pattern_len) == 0) {
		    char* remainder = str + i + pattern_len;
			memcpy(str + i, replace, replace_len);
			memmove(str + i + replace_len, remainder, strlen(remainder) + 1);
		}
```

lfittl commented on src/pg_query_summary_truncate.c:
```diff
@@ -0,0 +1,684 @@
+		if (memcmp(str + i, pattern, plen) == 0) {
+			size_t len = strlen(str + i + plen);
+			memmove(str + i + slen, str + i + plen, len + 1);
+		}
```
```suggestion
		if (memcmp(str + i, pattern, plen) == 0) {
		    char* remainder = str + i + plen;
			memmove(str + i + slen, remainder, strlen(remainder) + 1);
		}
```

duckinator committed:
[truncate] valuesList is a list of lists.

duckinator committed:
[truncate] Fix SELECT WHERE, and create a generic global_replace() function.

duckinator committed:
[truncate] Remove unintentionally-added script.

duckinator committed:
[truncate] I meant to move the T_InsertStmt case, not the IndexStmt case.

duckinator committed:
[truncate] MergeStmt will be handled after this PR.

duckinator committed:
[truncate] Make error messages more consistent.

duckinator committed:
[truncate] Fail immediately for unrecoverable errors; add helper functions for common patterns involved in this.

duckinator committed:
[truncate] remove debug prints

duckinator committed:
[truncate] significantly clean up apply_truncations

duckinator committed:
[truncate] whitespace cleanup

duckinator committed:
[truncate] Don't create truncations that would make it longer or keep it the same length.

lfittl committed:
Extract additional functions from Postgres source

lfittl committed:
pgindent run on modified files

lfittl committed:
Rework deparser use in summary truncation code

Previous commits added pg_query_deparse_opts, which acted like a
user-facing API in that it entered/exited the top memory context. However
the intent was to call this from the truncate code multiple times, which
needs a different memory context and error handling.

Instead, simplify and call deparseRawStmtOpts directly. This also avoids
duplicative elog(ERROR, ..) calls that are no longer necessary since the
deparser code itself can now jump to the top-level, where a new
PG_TRY/PG_CATCH block catches these errors.

lfittl committed:
Drop incorrect malloc use

Because we serialize the output as protobuf, we should not be using
malloc for the internal result (instead palloc should be used).

In passing, do an explicit pfree of the truncated_query for consistency.

duckinator commented on src/pg_query_summary.c:
```diff
@@ -619,11 +621,10 @@ pg_query_summary_walk(Summary * summary, Node *tree)
+	 * - We process cte_names as we iterate through the tree. - We only add
+	 * items to tables if they are *not* in cte_names. - CTEs can be defined
+	 * *after* they're referenced as names. - Because of this, we have to
+	 * store range_vars and process it after the initial tree walk.
```
This was messed up by the reformatting.

duckinator replied:
```suggestion
	 * NOTE regarding cte_names and range_vars:
	 * - We process cte_names as we iterate through the tree.
	 * - We only add items to tables if they are *not* in cte_names.
	 * - CTEs can be defined *after* they're referenced as names.
	 * - Due to this, we store range_vars and process it after the tree walk.
```

duckinator commented on src/pg_query_summary.c:
```diff
@@ -789,14 +775,14 @@ summary_to_protobuf(PgQuerySummaryParseResult * result, Summary * summary)
+	else
+	{
+		sr.truncated_query = summary->truncated_query;
+	}
```
Maybe this instead?

```suggestion
	if (summary->truncated_query == NULL)
		sr.truncated_query = pstrdup("");
	else
		sr.truncated_query = summary->truncated_query;
```

duckinator commented on test/summary_truncate.c:
```diff
@@ -0,0 +1,215 @@
+    char* query = "WITH \"原チコ氏にはす腹腹腹腹腹腹腹腹腹腹腹\" AS (SELECT) SELECT w".to_string();
+    query[0..=21].to_string();
+}
+*/
```
I'm not sure how to re-add this. I don't think the C test infrastructure supports an equivalent right now.

duckinator replied:
Actually, I'm not sure this test does anything useful. It seems like it's testing how Rust works, not implementation details?

seanlinsley replied:
Yes it was added at the same time as the test above it, documenting string truncation behavior in Rust: https://github.com/pganalyze/pg_query.rs/pull/7

duckinator replied:
Okay, so since we're truncating to a certain number of characters (not bytes) we don't have that problem. Thanks for confirming that. ^^

duckinator committed:
Fix comment messed up my pgindent.

duckinator committed:
[truncate] Remove commented-out Rust test that was just testing Rust functionality which has no C equivalent.

duckinator committed:
Remove unneeded curly braces around a few one-line if/else branches.

lfittl reviewed (approved):
:+1:
